README
# Lancet Data Extractions

## Data

The full database can be found in the `/data` folder.

### countries.csv

This file lists the country IDs used in the database and indicator files.
The IDs used are the ISO 3166 3-letter codes.

## Testing

To test the extractions, run `pytest`. It invokes tests defined in the
`/tests` folder.

## Caveats

Only countries are retained. Diamond Princess, for example, is not.


data/CHANGELOG.md
# Changelog

## December 10, 2020

- Add icl_i12_health_1 (% of people who are taking preventative measures: Worn a face mask outside your home (e.g. when on public transport, going to a supermarket, going to a main road))
- Add icl_i12_health_2 (% of people who are taking preventative measures: Washed hands with soap and water)
- Add icl_i12_health_3 (% of people who are taking preventative measures: Used hand sanitiser)
- Add icl_i12_health_4 (% of people who are taking preventative measures: Covered your nose and mouth when sneezing or coughing)
- Add icl_i12_health_5 (% of people who are taking preventative measures: Avoided contact with people who have symptoms or you think may have been exposed to the coronavirus)
- Add icl_i12_health_6 (% of people who are taking preventative measures: Avoided going out in general)
- Add icl_i12_health_7 (% of people who are taking preventative measures: Avoided going to hospital or other healthcare settings)
- Add icl_i12_health_8 (% of people who are taking preventative measures: Avoided taking public transport)
- Add icl_i12_health_9 (% of people who are taking preventative measures: Avoided working outside your home (if job is full or part-time))
- Add icl_i12_health_10 (% of people who are taking preventative measures: Avoided letting your children go to school/ university  (if you have children))
- Add icl_i12_health_11 (% of people who are taking preventative measures: Avoided having guests to your home)
- Add icl_i12_health_12 (% of people who are taking preventative measures: Avoided small social gatherings (not more than 2 people))
- Add icl_i12_health_13 (% of people who are taking preventative measures: Avoided medium-sized social gatherings (between 3 and 10 people))
- Add icl_i12_health_14 (% of people who are taking preventative measures: Avoided large-sized social gatherings (more than 10 people))
- Add icl_i12_health_15 (% of people who are taking preventative measures: Avoided crowded areas)
- Add icl_i12_health_16 (% of people who are taking preventative measures: Avoided going to shops)
- Add icl_i12_health_17 (% of people who are taking preventative measures: Slept in separate bedrooms at home, when normally you would share a bedroom)
- Add icl_i12_health_18 (% of people who are taking preventative measures: Eaten separately at home, when normally you would eat a meal with others)
- Add icl_i12_health_19 (% of people who are taking preventative measures: Cleaned frequently touched surfaces in the home (e.g. doorknobs, toilets, taps))
- Add icl_i12_health_20 (% of people who are taking preventative measures: Avoided touching objects in public (e.g. elevator buttons or doors))

## November 18, 2020

- Remove owid_weekly_excess_mortality_p_score_0_14 (The P-score (for ages 0-14) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019). Reason: This indicator is no longer provided by Our World in Data.

## October 27, 2020

- Add owid_weekly_excess_mortality_p_score_0_14 (The P-score (for ages 0-14) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)
- Add owid_weekly_excess_mortality_p_score_15_64 (The P-score (for ages 15-64) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)
- Add owid_weekly_excess_mortality_p_score_65_74 (The P-score (for ages 65-74) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)
- Add owid_weekly_excess_mortality_p_score_75_84 (The P-score (for ages 75-84) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)
- Add owid_weekly_excess_mortality_p_score_85plus (The P-score (for ages 85+) is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)
- Add owid_weekly_deaths_2020 (The number of weekly deaths, for all ages, from all causes in 2020)
- Add owid_weekly_deaths_2019 (The number of weekly deaths, for all ages, from all causes in 2019)
- Add owid_weekly_deaths_2018 (The number of weekly deaths, for all ages, from all causes in 2018)
- Add owid_weekly_deaths_2017 (The number of weekly deaths, for all ages, from all causes in 2017)
- Add owid_weekly_deaths_2016 (The number of weekly deaths, for all ages, from all causes in 2016)
- Add owid_weekly_deaths_2015 (The number of weekly deaths, for all ages, from all causes in 2015)
- Add owid_avg_weekly_deaths_2015_2019 (The average number of weekly deaths, for all ages, from all causes over the period 2015-2019)

## October 26, 2020

- Add owid_weekly_excess_mortality_p_score (The P-score is the percentage difference between the number of weekly deaths in 2020 and the average number of deaths in the same week over the years 2015–2019)

## October 23, 2020

- Add sdsn_positive_test_rate_smoothed (The share of COVID-19 tests that are positive, averaged over a rolling 7-day window)

## October 9, 2020

- Remove sdsn_positive_rate_smoothed (the indicator was based on owid_positive_rate, which is already a 7-day average)
- Remove sdsn_tests_per_case_smoothed (the indicator was based on owid_tests_per_case, which is already a 7-day average)

## October 7, 2020

- Rename sdsn_transmission_classification to sdsn_overall_transmission
- Rename sdsn_daily_cases_per_million_smoothed to sdsn_new_cases_per_million_smoothed
- Rename sdsn_daily_deaths_per_million_smoothed to sdsn_new_deaths_per_million_smoothed

## October 2, 2020

- Add sdsn_daily_cases_per_million_smoothed (Daily new confirmed cases of COVID-19 per 1,000,000 people, averaged over a rolling 14-day window)

pipelines/yougov.py
import requests
import pandas
import demjson
from datetime import datetime
from time import gmtime, strftime

from helpers.normalize_date import normalize_date
from helpers.save_indicator import save_indicator
from helpers.normalize_country import normalize_country

# This request replicates the request made by the browser for the various
# charts displayed on the YouGov website.
url = "https://chartserver.live/server/yg/v4/data.js?x=987984"
request = requests.get(url)

# Convert the JS object contained in the response into a python dictionary
text = request.text
js_object = text[text.index("{") : text.rindex("}") + 1]
data_dict = demjson.decode(js_object)

# Map our indicator IDs to the title of the YouGov indicator
INDICATOR_MAP = {
    "fear_of_catching": "#1 YouGov COVID-19 tracker: fear of catching",
    "government_handling": "#2 YouGov COVID-19 tracker: government handling",
    "avoiding_crowded_places": "#3 YouGov COVID-19 behaviour changes tracker: Avoiding crowded public places",
    "wearing_mask_in_public": "#4 YouGov COVID-19 behaviour changes tracker: Wearing a face mask when in public places ",
    "avoiding_going_to_work": "#5 YouGov COVID-19 behaviour changes tracker: Avoiding going to work",
    "avoiding_raw_meat": "#6 YouGov COVID-19 behaviour changes tracker: Avoiding raw meat",
    "stopping_sending_children": "#7 YouGov COVID-19 behaviour changes tracker: Stopping sending children to child care or school",
    "improving_personal_hygiene": "#8 YouGov COVID-19 behaviour changes tracker: Improving personal hygiene",
    "refraining_from_touching_objects": "#9 YouGov COVID-19 behaviour changes tracker: Refraining from touching objects in public",
    "avoiding_contact_with_tourists": "#10 YouGov COVID-19 behaviour changes tracker: Avoiding physical contact with tourists",
    "support_stopping_flights_from_china": "#11 YouGov COVID-19 measures supported tracker: Stopping all flights coming into country from mainland China",
    "support_quarantine_flights_from_china": "#12 YouGov COVID-19 measures supported tracker: Quarantining all passengers on all flights coming into country from mainland China ",
    "support_stopping_international_flights": "#13 YouGov COVID-19 measures supported tracker: Stopping all inbound international flights from countries with confirmed cases of coronavirus",
    "support_quarantine_international_flights": "#14 YouGov COVID-19 measures supported tracker: Quarantining all inbound international flights from countries with confirmed cases of coronavirus",
    "support_quarantine_chinese_travellers": "#15 YouGov COVID-19 measures supported tracker: Quarantining all Chinese travellers currently in country",
    "support_quarantine_any_person": "#16 YouGov COVID-19 measures supported tracker: Quarantining anyone who has been in contact with a contaminated patient ",
    "support_quarantine_any_location": "#17 YouGov COVID-19 measures supported tracker: Quarantining any location in country that a contaminated patient has been in",
    "support_free_masks": "#18 YouGov COVID-19 measures supported tracker: Providing free masks for all people in country",
    "support_work_from_home": "#19 YouGov COVID-19 measures supported tracker: encourage working from home",
    "support_temporarily_close_schools": "#20 YouGov COVID-19 measures supported tracker: temporarily close schools",
    "support_cancel_large_events": "#21 YouGov COVID-19 measures supported tracker: cancel large events",
    "support_cancel_hospital_routines": "#22 YouGov COVID-19 measures supported tracker: cancel routine hospital procedures",
    "confidence_in_health_authorities": "#23 YouGov COVID-19 tracker: confidence in health authorities",
    "perceived_national_improvement": "#24 YouGov COVID-19 tracker: perceived national improvement",
    "perceived_global_improvement": "#25 YouGov COVID-19 tracker: perceived global improvement",
    "international_happiness": "#26 YouGov COVID-19 tracker: international happiness",
    # The following indicator is not shown on the YouGov website (#23 is shown
    # instead) and is thus not used:
    # 'confidence_in_health_authorities': '#27 YouGov COVID-19 tracker: confidence in health authorities',
    "personal_health_fears": "#28 YouGov COVID-19 tracker: personal health fears",
    "friends_and_family_health_fears": "#29 YouGov COVID-19 tracker: friends and family health fears",
    "finances_fears": "#30 YouGov COVID-19 tracker: finances fears",
    "job_loss_fears": "#31 YouGov COVID-19 tracker: job loss fears",
    "education_fears": "#32 YouGov COVID-19 tracker: education fears",
    "social_impact_fears": "#33 YouGov COVID-19 tracker: social impact fears",
    # The following two indicators are currently not being shown on the YouGov website
    # and are thus not used:
    # 'support_stopping_all_flights': '#34 YouGov COVID-19 measures supported tracker: stopping all inbound flights',
    # 'support_quarantine_all_flights': '#35 YouGov COVID-19 measures supported tracker: quarantining all inbound airline passengers',
}


def run_pipeline(indicator):
    # Find the chart series for this indicator
    chart_series = None
    # The YouGov label to look for
    needle = INDICATOR_MAP[indicator.replace("yougov_", "")]
    for id, object in data_dict.items():
        label = id.replace("chart_", "#") + " " + object["title"]
        if needle.lower() == label.lower():
            chart_series = object["chartSeries"]
            break

    # Convert the data in the dict into a pandas dataframe:
    # The series contains the country name and a data object
    # The data object is a list of data points in the format [timestamp, value]
    data = []
    for series in chart_series:
        for observation in series["data"]:
            data.append(
                {
                    "country": series["name"],
                    "date": strftime("%Y-%m-%d", gmtime(observation[0] / 1000)),
                    "timestamp": observation[0],
                    "value": observation[1],
                }
            )

    # Convert dict to dataframe
    dataset = pandas.DataFrame.from_dict(data)

    # Normalize countries
    dataset["iso_code"] = dataset["country"].apply(
        lambda country: normalize_country(country)
    )

    # Normalize date
    dataset["date"] = dataset["date"].apply(
        lambda date: normalize_date(date, "%Y-%m-%d")
    )

    # Rename the value column
    dataset = dataset.rename(columns={"value": indicator})

    # Temporary: yougov_support_cancel_hospital_routines has some negative
    # timestamp values. Those seem like a mistake, so we drop them.
    if indicator == "yougov_support_cancel_hospital_routines":
        print("Dropping negative timestamp")
        dataset = dataset[dataset.timestamp != -500000]

    # Create slice of data with country ID, date, and indicator
    dataset = dataset[["iso_code", "date", indicator]]

    save_indicator(indicator, dataset=dataset)
